kernel/net/src/interface.rs

```diff
@@ -0,0 +1,67 @@
+use crate::{device::DeviceWrapper, Device, Result};
+use alloc::{collections::BTreeMap, sync::Arc};
+use irq_safety::MutexIrqSafe;
+use smoltcp::{iface, phy::DeviceCapabilities, wire};
+
+pub use smoltcp::iface::SocketSet;
+pub use wire::{IpAddress, IpCidr};
+
+#[derive(Clone)]
+pub struct Interface {
+    // FIXME: Can this be a regular mutex?
+    inner: Arc<MutexIrqSafe<iface::Interface<'static>>>,
```

Comment:
if/when we use deferred interrupt handlers here, then yes, we'd be able to partition the NIC type into things that need to be accessed from the interrupt handler and things that only need to be accessed in the deferred handler (or any other normal execution context). 

But for now it needs to be wrapped in this because it's jointly accessed from the NIC's interrupt handler.

Reply:
oh wait on second thought, this doesn't appear to be a type that accesses any NIC-specific h/w stuff during an interrupt handler. 
Hmm, what type was this derived from in the existing main branch?

Reply:
I think this is equivalent to `smoltcp::EthernetInterface<'_, '_, '_, EthernetDevice<N>>` of `EthernetNetworkInterface.iface` in `ethernet_smoltcp_device`. The new versions of `smoltcp` separate the interface and device. The interface doesn't store the device, but polling it requires a reference to the device.

Reply:
Interesting! Since I'm not intimately familiar with the new smoltcp design, can you state with certainty that this is never accessed or accessible from any NIC interrupt handler (just e1000 for now i guess)?  If so, meaning that it's only ever accessed from a regular execution (non-interrupt handler) context, then yes we can reduce this to a regular `Mutex` instead of an interrupt-safe one.

Reply:
Yep. The implementor crates only access the concrete device type e.g.:
```rust
static E1000_NIC: Once<MutexIrqSafe<E1000Nic>> = Once::new();
```

The interface is a layer on top of the device that the interrupt handler never needs to touch.

I changed it to use a `MutexSleep` in [d9ab2e7](https://github.com/theseus-os/Theseus/pull/677/commits/d9ab2e75c0d23c70dc2d10cd2f86e3e324369b40).